newIDE/app/src/CommandPalette/CommandManager.js

```diff
@@ -0,0 +1,103 @@
+// @flow
+import { type MessageDescriptor } from '../Utils/i18n/MessageDescriptor.flow';
+
+type CommandHandler = () => void | Promise<void>;
+
+export type Command = {|
+  displayText: MessageDescriptor,
+  enabled: boolean,
+  handler: CommandHandler,
+|};
+
+export type NamedCommand = {|
+  name: string,
+  ...Command,
+|};
+
+export type CompoundCommandOption<T> = {|
+  value: T,
+  handler: CommandHandler,
+  text: string,
+  iconSrc?: string,
+|};
+
+export type CompoundCommand<T> = {|
+  displayText: MessageDescriptor,
+  enabled: boolean,
+  generateOptions: () => Array<CompoundCommandOption<T>>,
+|};
```

Comment:
Two questions:
* Could we simply rename `CompoundCommand` to `CommandWithOptions`. Compound seems a bit complex for someone that do not have the context, and after all it's no more and no less than a command that has a bunch of options :)
* Going a bit further, could we have a single `Command` type, but with an optional `generateOptions`? (`generateOptions?: () => ...`)

Trying to understand if we really need to make the distinction between these.

Reply:
Note that in Flow you can also do something like:

```
export type Command = {|
  displayText: MessageDescriptor,
  enabled: boolean,
  handler: CommandHandler,
|} | {|
  displayText: MessageDescriptor,
  enabled: boolean,
  generateOptions: () => Array<CompoundCommandOption<T>>,
|}
```

Though this can somewhat leads to complex error messages. But basically this tells Flow that a command is either the first shape or the other, so it can have handler OR a generateOptions but not both - and it will force you to check this when needed.

Reply:
Okay, that's what I was about to suggest, in place of an optional `generateOptions` :D 
I'll do this right away!

Reply:
Did this, and removed the redundancy from the command manager functions. I couldn't merge the hooks into one, though. I tried something like this:

```js
// Attempted unified hook:
const useCommand = (commandName: string, command: Command) => {
  const {displayText, enabled} = command;
  React.useEffect(() => {
    if (!enabled) return;
    if (command.handler) {
      // Direct command
      commandManager.registerCommand(commandName, {
        displayText, enabled, handler: command.handler
      })
    } else {
      // Command with options
      commandManager.registerCommand(commandName, {
        displayText, enabled, generateOptions: command.generateOptions
      })
    }
  }, [commandManager, commandName, displayText, enabled, command.handler, command.generateOptions])
}
```

The issue is that we get stuck between React warnings and Flow errors in the above code. Flow won't allow us to add `command.handler` or `command.generateOptions` to the dependency list because each of those properties are absent in one or the other of the enum types of `Command`, while omitting any of those from the dependency list or using React hooks in not the top-level of custom hook is not right.